rpc/jsonrpc/types/types.go

```diff
@@ -94,6 +124,55 @@ func (req RPCRequest) String() string {
 	return fmt.Sprintf("RPCRequest{%s %s/%X}", req.ID, req.Method, req.Params)
 }
 
+// MakeResponse constructs a success response to req with the given result.  If
+// there is an error marshaling result to JSON, it returns an error response.
+func (req RPCRequest) MakeResponse(result interface{}) RPCResponse {
+	data, err := json.Marshal(result)
+	if err != nil {
+		return req.MakeErrorf(CodeInternalError, "marshaling result: %v", err)
+	}
+	return RPCResponse{ID: req.ID, Result: data}
+}
+
+// MakeErrorf constructs an error response to req with the given code and a
+// message constructed by formatting msg with args.
+func (req RPCRequest) MakeErrorf(code ErrorCode, msg string, args ...interface{}) RPCResponse {
+	return RPCResponse{
+		ID: req.ID,
+		Error: &RPCError{
+			Code:    int(code),
+			Message: code.String(),
+			Data:    fmt.Sprintf(msg, args...),
+		},
+	}
+}
+
+// MakeError constructs an error response to req from the given error value.
+// This function will panic if err == nil.
+func (req RPCRequest) MakeError(err error) RPCResponse {
+	if err == nil {
+		panic("cannot construct an error response for nil")
```

Comment:
why not just return an internal error for this?

Reply:
I'm not sure it makes sense to ship this case out to a client. There's no error value to derive a message string from, and it means the server code is statically wrong (i.e., there's nothing about the specific request—it's wrong for all possible paths).

Reply:
(also note that in this case the RPC handler will already log the panic and report an error back to the caller, so it won't actually crash the node—though I think it _ought_ to)